src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -133,7 +133,20 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
     protected ComponentState GetComponentState(int componentId)
         => GetRequiredComponentState(componentId);
 
-    internal ComponentState GetComponentState(IComponent component)
+    /// <summary>
+    /// Gets the <see cref="IComponentRenderMode"/> for a given component if available.
+    /// </summary>
+    /// <param name="component">The component type</param>
+    /// <returns></returns>
+    protected internal virtual IComponentRenderMode? GetComponentRenderMode(IComponent component)
+        => null;
```

Comment:
I'm a bit torn about this as it's possible to avoid exposing this new API (e.g., with the "filter callback" I prototyped) but at the same time I don't have a really objective reason to say it's important to avoid exposing it.

It looks like the tradeoff is:

 * With filter callback:
   * (Good) No need have a renderer-level API that returns component rendermodes
   * (Arguably good) The exact rules about resolving "registration rendermode" and "component rendermode" are delegated to the renderer, so each renderer could have its own semantics as required
   * (Bad) The API with a filter callback is less obvious
 * With `GetComponentRenderMode` and other aspects of how it's done in this PR:
   * The opposite, i.e., the API is a bit more obvious but at the cost that we need to lock all renderers to being able to say at any moment what is the rendermode for a given component, and we can not allow renders to vary anything about the semantics of how "registration rendermode" and "component rendermode" are combined.

Do you agree that's the tradeoff? I'm not trying to demand a change here since both seem like valid approaches. I just want to be sure I've understood and check if you think there are other subtleties or if it might be worth going to the "filter callback" approach (would be a pretty easy change within this PR).

Reply:
> The opposite, i.e., the API is a bit more obvious but at the cost that we need to lock all renderers to being able to say at any moment what is the rendermode for a given component, and we can not allow renders to vary anything about the semantics of how "registration rendermode" and "component rendermode" are combined.

It's not mandatory for a renderer to provide a RenderMode, in fact, the endpoint HTML renderer is the only one that provide is it.

We could remove that from the public API of the renderer and instead pass it as a callback to `PersistStateAsync` but I didn't want to add an additional parameter to it as that required a new overload and potential breaking change.

From my view, you only implement `GetComponentRenderMode` if your renderer handles multiple render modes. Otherwise, the default implementation is fine. As in any case, any caller will have to account for the component not having a render mode.

Alternatively, we could also put this API in `ComponentState` which would mean something like `Renderer.GetComponentState(component).GetRenderMode()`

Reply:
IComponentRenderMode permeates through the core `Microsoft.AspNetCore.Components` API, so I think it's fine having methods that interact with the abstraction as long as we don't bake in behaviors specific to the concrete set of render modes we provide.